### ISSUE TITLE
mapfishapp print - issue with scale

### ISSUE THREAD

**Issue by fvanderbiest** [georchestra/georchestra#1084]
Basel / Strasbourg is ~ 114 km, see http://fr.distance.to/Strasbourg/Basel

OK:
![0b7ebece-5638-11e5-972a-9eeb9d8a8586](https://cloud.githubusercontent.com/assets/265319/10427337/172d5890-70e9-11e5-9a45-84b804a655c5.png)

NOK: more than 200km
![07c10b70-5638-11e5-87eb-b31636833243](https://cloud.githubusercontent.com/assets/265319/10427340/1fa2a66a-70e9-11e5-970e-b5f377f52bd0.png)

**Comment by fvanderbiest** [comment]
Might be related to https://github.com/mapfish/mapfish-print/issues/264 or https://github.com/mapfish/mapfish-print/issues/327 (thanks @jusabatier for the findings)

**Comment by fvanderbiest** [comment]
"geodetic" patch from mapfish print v3 is to be backported to 1.2.x (not so easy !)

**Comment by fvanderbiest** [comment]
Nice to have with this fix: #985 solved too.

**Comment by MaelREBOUX** [comment]
Happy to see work on this issue.

Some users reported a real issue with printing. The map-image is not a the scale when compare to the graphic scale. I think it is the same issue ?

When printing a PDF with "real scale" option in Acrobat Reader we see that the graphic scale is correct but absolutely not the map content. The distorsion is about 20-30 % at 1:200.

We have many many users that make mesures (that they should not do...) from the paper print.

correct graphic scale

![screenshot519](https://cloud.githubusercontent.com/assets/6370443/22256608/9699df74-e25b-11e6-9cd8-16be5665dbbf.png)

incorrect map size

![screenshot520](https://cloud.githubusercontent.com/assets/6370443/22256626/a201716a-e25b-11e6-9207-cdbc91f4ec0c.png)

**Comment by fvanderbiest** [comment]
> I think it is the same issue ?

Correct.

**Comment by Vampouille** [comment]
Post sur /mapfishapp/pdf/create.json?url=/mapfishapp/pdf
Post data : 

```json 
{
   "units":"m",
   "srs":"EPSG:3857",
   "layout":"A4 portrait",
   "dpi":91,
   "outputFormat":"pdf",
   "showLegend":true,
   "outputFilename":"georchestra_${yyyy-MM-dd_hhmmss}",
   "layers":[
      {
         "minScaleDenominator":266.5911979812228,
         "maxScaleDenominator":8735660.375448719,
         "baseURL":"https://www.cigalsace.org/geoserver/cigal/wms?SERVICE=WMS&",
         "opacity":1,
         "type":"WMS",
         "version":"1.1.1",
         "layers":[
            "CIGAL_LignesFrontieres_250000_UTM32"
         ],
         "format":"image/png",
         "styles":[
            ""
         ],
         "singleTile":false,
         "customParams":{
            "TRANSPARENT":"true",
            "EXCEPTIONS":"application/vnd.ogc.se_xml",
            "SLD_VERSION":"1.0.0"
         }
      },
      {
         "minScaleDenominator":266.5911979812228,
         "maxScaleDenominator":559082264.028718,
         "baseURL":"https://www.cigalsace.org/geoserver/cigal/ows?SERVICE=WMS&",
         "opacity":1,
         "type":"WMS",
         "version":"1.3.0",
         "layers":[
            "CIGAL_SCAN25_2006_ALSACE_TIFA_CC48"
         ],
         "format":"image/png",
         "styles":[
            ""
         ],
         "singleTile":false,
         "customParams":{
            "TRANSPARENT":"true",
            "EXCEPTIONS":"XML",
            "SLD_VERSION":"1.1.0"
         }
      }
   ],
   "pages":[
      {
         "center":[
            824669.65262971,
            6128078.8154104
         ],
         "scale":1500000,
         "rotation":0,
         "mapTitle":"",
         "mapComments":"",
         "copyright":"Source : Source : © IGN SCAN 25 2005 Paris (Partenariat CIGAL - www.cigalsace.org)",
         "scaleLbl":"Échelle : ",
         "dateLbl":"Date : ",
         "showOverview":true,
         "showNorth":true,
         "showScalebar":true,
         "showDate":true,
         "showLegend":true,
         "projection":"Projection : Web Spherical Mercator"
      }
   ],
   "legends":[
      {
         "name":"CIGAL SCAN25 2006",
         "classes":[
            {
               "name":"",
               "icons":[
                  "https://www.cigalsace.org/geoserver/cigal/ows?SERVICE=WMS&TRANSPARENT=true&VERSION=1.3.0&EXCEPTIONS=application%2Fvnd.ogc.se_xml&REQUEST=GetLegendGraphic&SLD_VERSION=1.1.0&CRS=EPSG%3A3857&LAYER=CIGAL_SCAN25_2006_ALSACE_TIFA_CC48&FORMAT=image%2Fpng&LEGEND_OPTIONS=forceLabels%3Aon%3BfontAntiAliasing%3Atrue&SCALE=1500000"
               ]
            }
         ]
      },
      {
         "name":"CIGAL limites",
         "classes":[
            {
               "name":"",
               "icons":[
                  "https://www.cigalsace.org/geoserver/cigal/wms?SERVICE=WMS&TRANSPARENT=true&VERSION=1.1.1&REQUEST=GetLegendGraphic&EXCEPTIONS=application%2Fvnd.ogc.se_xml&SLD_VERSION=1.0.0&LAYER=CIGAL_LignesFrontieres_250000_UTM32&FORMAT=image%2Fpng&LEGEND_OPTIONS=forceLabels%3Aon%3BfontAntiAliasing%3Atrue&SCALE=1500000"
               ]
            }
         ]
      }
   ]
}
```

**Comment by Vampouille** [comment]
We need to add this : https://github.com/mapfish/mapfish-print/pull/286/files#diff-4b012a3744bd5718903747fd88fb042bR110

**Comment by Vampouille** [comment]
here : 
https://github.com/mapfish/mapfish-print/blob/daeea96d2d3b33d6ccf02fe62d81b9c7aa2574a2/src/main/java/org/mapfish/print/config/layout/ScalebarBlock.java#L90

**Comment by fvanderbiest** [comment]
bumping ...

**Comment by Vampouille** [comment]
Upstream PR : https://github.com/mapfish/mapfish-print/pull/497
For geOrchestra : https://github.com/Vampouille/mapfish-print/commit/a582798a3f7cceb567ad9ad57c84ec94d46085fc
And branch : https://github.com/Vampouille/mapfish-print/tree/2.1.x-GEORCHESTRA-geodetic

**Comment by Vampouille** [comment]
Commit for new version of mapfish-print https://github.com/georchestra/georchestra/commit/beffa37403852bbc5b3b85f5ca384d2bb277087a (2.1-GEORCHESTRA-geodetic)

**Comment by Vampouille** [comment]
Waiting for installation of mapfish-print (2.1-GEORCHESTRA-geodetic) on packages.georchestra.org

**Comment by pmauduit** [comment]
done: https://packages.georchestra.org/maven/repository/org/mapfish/print/print-lib/2.1-GEORCHESTRA-geodetic/

**Comment by pmauduit** [comment]
Travis & community CI should pass on the sandbox branch (see commit above)

**Comment by fvanderbiest** [comment]
sandbox ... is it automatically deployed somewhere ?

**Comment by fvanderbiest** [comment]
Tested locally with jetty, the results are correct. Good job !

**Comment by fvanderbiest** [comment]
Backported to 15.12 with https://github.com/georchestra/georchestra/commit/66d7fe070f355650ec6702b9dcf0a6505131c254

**Comment by jusabatier** [comment]
I rebuild MFapp with "2.1-GEORCHESTRA-geodetic" and deploy the new version on a test platform, but the generated image still false when printed.

I have a correct graphic scale but the image printed isn't at the requested scale.

In order to test this, I tried on the center round of a soccer field which is 9,15m.
Whith a scale of 1/200, it measure 6,5cm on the printed map, what is 13m in real.
It should measure ~4,6cm to be correct.

So I don't think this issue is solved yet or I miss something in my tests...

**Comment by Vampouille** [comment]
We also need to add `geodetic` parameter on JSON spec. Print with geodetic parameter 
[local_1_800000_geodetic.pdf](https://github.com/georchestra/georchestra/files/776728/local_1_800000_geodetic.pdf)
With a scale at 1:80000 :

* 25km scale bar : 3.1cm
* 97km between Strasbourg and Mulhouse : 12.1cm

**Comment by fvanderbiest** [comment]
Right, see http://www.mapfish.org/doc/print/protocol.html#print-pdf

**Comment by Vampouille** [comment]
Same spec without `geodetic` parameter : 
[local_1_800000.pdf](https://github.com/georchestra/georchestra/files/776731/local_1_800000.pdf)


Browser display wrong bbox because it don't use geodetic correction :
![spec2](https://cloud.githubusercontent.com/assets/7067560/22970378/3dc8935e-f372-11e6-814d-305b39a9b535.png)

`geodetic` parameter need to be added at top level : 

```json
{
  "units":"m",
  "srs":"EPSG:3857",
  "layout":"A4 portrait",
  "dpi":91,
  "outputFormat":"pdf",
  "showLegend":true,
  "geodetic": true,
  "outputFilename":"georchestra_${yyyy-MM-dd_hhmmss}",
....
```

**Comment by Vampouille** [comment]
Full working spec : 

```json
{
  "units":"m",
  "srs":"EPSG:3857",
  "layout":"A4 portrait",
  "dpi":91,
  "outputFormat":"pdf",
  "showLegend":true,
  "geodetic": true,
  "outputFilename":"georchestra_${yyyy-MM-dd_hhmmss}",
  "layers":[
    {
      "minScaleDenominator":266.5911979812228,
      "maxScaleDenominator":559082264.028718,
      "baseURL":"https://www.cigalsace.org/geoserver/cigal/ows?SERVICE=WMS&",
      "opacity":1,
      "type":"WMS",
      "version":"1.3.0",
      "layers":[
        "CIGAL_SCAN25_2006_ALSACE_ECW_CC48"
      ],
      "format":"image/png",
      "styles":[
        ""
      ],
      "singleTile":false,
      "customParams":{
        "TRANSPARENT":"true",
        "EXCEPTIONS":"XML",
        "SLD_VERSION":"1.1.0"
      }
    },
    {
      "minScaleDenominator":266.5911979812228,
      "maxScaleDenominator":559082264.028718,
      "baseURL":"https://www.cigalsace.org/geoserver/CG67/ows?SERVICE=WMS&",
      "opacity":1,
      "type":"WMS",
      "version":"1.3.0",
      "layers":[
        "CG67_SCAN25_BR_2010_TIF_CC48"
      ],
      "format":"image/png",
      "styles":[
        ""
      ],
      "singleTile":false,
      "customParams":{
        "TRANSPARENT":"true",
        "EXCEPTIONS":"XML",
        "SLD_VERSION":"1.1.0"
      }
    },
    {
      "minScaleDenominator":266.5911979812228,
      "maxScaleDenominator":150000,
      "baseURL":"https://www.cigalsace.org/geoserver/IGN/ows?SERVICE=WMS&",
      "opacity":1,
      "type":"WMS",
      "version":"1.3.0",
      "layers":[
        "IGN_bdTopo_commune"
      ],
      "format":"image/png",
      "styles":[
        ""
      ],
      "singleTile":false,
      "customParams":{
        "TRANSPARENT":"true",
        "EXCEPTIONS":"XML",
        "SLD_VERSION":"1.1.0"
      }
    },
    {
      "minScaleDenominator":266.5911979812228,
      "maxScaleDenominator":559082264.028718,
      "baseURL":"https://www.cigalsace.org/geoserver/IGN/ows?SERVICE=WMS&",
      "opacity":1,
      "type":"WMS",
      "version":"1.3.0",
      "layers":[
        "IGN_bdTopo_hydronyme"
      ],
      "format":"image/png",
      "styles":[
        ""
      ],
      "singleTile":false,
      "customParams":{
        "TRANSPARENT":"true",
        "EXCEPTIONS":"XML",
        "SLD_VERSION":"1.1.0"
      }
    },
    {
      "minScaleDenominator":266.5911979812228,
      "maxScaleDenominator":559082264.028718,
      "baseURL":"https://www.cigalsace.org/geoserver/IGN/ows?SERVICE=WMS&",
      "opacity":1,
      "type":"WMS",
      "version":"1.3.0",
      "layers":[
        "IGN_bdTopo_surface_eau"
      ],
      "format":"image/png",
      "styles":[
        ""
      ],
      "singleTile":false,
      "customParams":{
        "TRANSPARENT":"true",
        "EXCEPTIONS":"XML",
        "SLD_VERSION":"1.1.0"
      }
    },
    {
      "baseURL":"https://www.cigalsace.org/geoserver/REF/ows?SERVICE=WMS&",
      "opacity":0.44,
      "type":"WMS",
      "version":"1.3.0",
      "layers":[
        "IGN_SCAN1000_TIF_L93"
      ],
      "format":"image/png",
      "styles":[
        ""
      ],
      "singleTile":false,
      "customParams":{
        "EXCEPTIONS":"XML",
        "TRANSPARENT":true,
        "SLD_VERSION":"1.1.0"
      }
    },
    {
      "minScaleDenominator":266.5911979812228,
      "maxScaleDenominator":8735660.375448719,
      "baseURL":"https://www.cigalsace.org/geoserver/cigal/wms?SERVICE=WMS&",
      "opacity":1,
      "type":"WMS",
      "version":"1.1.1",
      "layers":[
        "CIGAL_LignesFrontieres_250000_UTM32"
      ],
      "format":"image/png",
      "styles":[
        ""
      ],
      "singleTile":false,
      "customParams":{
        "TRANSPARENT":"true",
        "EXCEPTIONS":"application/vnd.ogc.se_xml",
        "SLD_VERSION":"1.0.0"
      }
    }
  ],
  "pages":[
    {
      "center":[
        873017.33980169,
        6132112.7618516
      ],
      "scale":800000,
      "rotation":0,
      "mapTitle":"",
      "mapComments":"",
      "copyright":"Sources : © IGN SCAN 25 2005 Paris (Partenariat CIGAL - www.cigalsace.org), © IGN SCAN 25 2009, © IGN SCAN1000 2015",
      "scaleLbl":"Échelle : ",
      "dateLbl":"Date : ",
      "showOverview":true,
      "showNorth":true,
      "showScalebar":true,
      "showDate":true,
      "showLegend":true,
      "projection":"Projection : Web Spherical Mercator"
    }
  ],
  "legends":[
    {
      "name":"CIGAL limites",
      "classes":[
        {
          "name":"",
          "icons":[
            "https://www.cigalsace.org/geoserver/cigal/wms?SERVICE=WMS&TRANSPARENT=true&VERSION=1.1.1&REQUEST=GetLegendGraphic&EXCEPTIONS=application%2Fvnd.ogc.se_xml&SLD_VERSION=1.0.0&LAYER=CIGAL_LignesFrontieres_250000_UTM32&FORMAT=image%2Fpng&LEGEND_OPTIONS=forceLabels%3Aon%3BfontAntiAliasing%3Atrue&SCALE=800000"
          ]
        }
      ]
    },
    {
      "name":"IGN SCAN1000 France",
      "classes":[
        {
          "name":"",
          "icons":[
            "https://www.cigalsace.org/geoserver/REF/ows?SERVICE=WMS&EXCEPTIONS=application%2Fvnd.ogc.se_xml&TRANSPARENT=TRUE&VERSION=1.3.0&SLD_VERSION=1.1.0&REQUEST=GetLegendGraphic&CRS=EPSG%3A3857&LAYER=IGN_SCAN1000_TIF_L93&FORMAT=image%2Fpng&LEGEND_OPTIONS=forceLabels%3Aon%3BfontAntiAliasing%3Atrue&SCALE=800000"
          ]
        }
      ]
    },
    {
      "name":"IGN_bdTopo_surface_eau",
      "classes":[
        {
          "name":"",
          "icons":[
            "https://www.cigalsace.org/geoserver/IGN/ows?SERVICE=WMS&TRANSPARENT=true&VERSION=1.3.0&EXCEPTIONS=application%2Fvnd.ogc.se_xml&REQUEST=GetLegendGraphic&SLD_VERSION=1.1.0&CRS=EPSG%3A3857&LAYER=IGN_bdTopo_surface_eau&FORMAT=image%2Fpng&LEGEND_OPTIONS=forceLabels%3Aon%3BfontAntiAliasing%3Atrue&SCALE=800000"
          ]
        }
      ]
    },
    {
      "name":"IGN_bdTopo_hydronyme",
      "classes":[
        {
          "name":"",
          "icons":[
            "https://www.cigalsace.org/geoserver/IGN/ows?SERVICE=WMS&TRANSPARENT=true&VERSION=1.3.0&EXCEPTIONS=application%2Fvnd.ogc.se_xml&REQUEST=GetLegendGraphic&SLD_VERSION=1.1.0&CRS=EPSG%3A3857&LAYER=IGN_bdTopo_hydronyme&FORMAT=image%2Fpng&LEGEND_OPTIONS=forceLabels%3Aon%3BfontAntiAliasing%3Atrue&SCALE=800000"
          ]
        }
      ]
    },
    {
      "name":"CG67 SCAN25 2010",
      "classes":[
        {
          "name":"",
          "icons":[
            "https://www.cigalsace.org/geoserver/CG67/ows?SERVICE=WMS&TRANSPARENT=true&VERSION=1.3.0&EXCEPTIONS=application%2Fvnd.ogc.se_xml&REQUEST=GetLegendGraphic&SLD_VERSION=1.1.0&CRS=EPSG%3A3857&LAYER=CG67_SCAN25_BR_2010_TIF_CC48&FORMAT=image%2Fpng&LEGEND_OPTIONS=forceLabels%3Aon%3BfontAntiAliasing%3Atrue&SCALE=800000"
          ]
        }
      ]
    },
    {
      "name":"CIGAL SCAN25 2006",
      "classes":[
        {
          "name":"",
          "icons":[
            "https://www.cigalsace.org/geoserver/cigal/ows?SERVICE=WMS&TRANSPARENT=true&VERSION=1.3.0&EXCEPTIONS=application%2Fvnd.ogc.se_xml&REQUEST=GetLegendGraphic&SLD_VERSION=1.1.0&CRS=EPSG%3A3857&LAYER=CIGAL_SCAN25_2006_ALSACE_ECW_CC48&FORMAT=image%2Fpng&LEGEND_OPTIONS=forceLabels%3Aon%3BfontAntiAliasing%3Atrue&SCALE=800000"
          ]
        }
      ]
    }
  ]
}
```

**Comment by Vampouille** [comment]
@fvanderbiest Maybe we can add `geodetic` here : https://github.com/georchestra/georchestra/blob/15.12/mapfishapp/src/main/webapp/app/js/GEOR_print.js#L120

**Comment by fvanderbiest** [comment]
Yep, but the comment from the mapfish print page makes me doubt about the way we implemented it [elsewhere](https://github.com/georchestra/georchestra/blob/15.12/mapfishapp/src/main/webapp/app/addons/measurements/js/main.js#L39)

>The geodetic parameter can be set to true so the scale of geodetic layers can correctly be calculated. Certain projections (Google and Latlong for example) are based on a spheroid and therefore require 
> geodetic: true in order to correctly calculate the scale. If the geodetic parameter is not present it will be assumed to be false.

**Comment by fvanderbiest** [comment]
Collecting information from OpenLayers sources:

>The recommended geodesic setting for maps in EPSG:4326 is false, and true for EPSG:900913. If set to true, the scale will be calculated based on the horizontal size of the pixel in the center of the map viewport.

```
    /**
     * APIProperty: geodesic
     * {Boolean} Calculate geodesic metrics instead of planar metrics.  This
     *     requires that geometries can be transformed into Geographic/WGS84
     *     (if that is not already the map projection).  Default is false.
     */
```

>The metrics returned by the measure control are planar measures by
>default.  If your map is in a geographic projection or you have the
>appropriate projection definitions to transform your geometries into
>geographic coordinates, you can set the "geodesic" property of the control
>to true to calculate geodesic measures instead of planar measures.

**Comment by fvanderbiest** [comment]
Based on the above, commit e8dcf3c should yield a more accurate PDF.

**Comment by fvanderbiest** [comment]
The problem is now that the printed bbox does not match the one that is drawn by the client side.

![capture du 2017-02-15 12-58-58](https://cloud.githubusercontent.com/assets/265319/22973587/939d0cae-f37e-11e6-8095-1e1b497bcdc1.png)
yields:
![capture du 2017-02-15 12-59-15](https://cloud.githubusercontent.com/assets/265319/22973592/993353b2-f37e-11e6-87e3-ef9c60268528.png)

**Comment by fvanderbiest** [comment]
We need to work on https://github.com/georchestra/geoext/blob/master/lib/GeoExt/data/PrintPage.js#L305-L321

**Comment by fvanderbiest** [comment]
Patch attempt in https://gist.github.com/fvanderbiest/47dca09a67e3e8534e5e6ab60464b656#file-printpage-js-L324-L347

**Comment by Vampouille** [comment]
With @fvanderbiest patch and without mine, it works :

[georchestra_2017-02-16_010120.pdf](https://github.com/georchestra/georchestra/files/780020/georchestra_2017-02-16_010120.pdf)

See https://github.com/georchestra/georchestra/pull/1668 for revert of my commit.

**Comment by fvanderbiest** [comment]
With 7e2fda4 the bug should be fixed from 16.12 on.
